### ISSUE TITLE
Training only a specific Layer

### ISSUE THREAD

**Issue by abhigarg-iitk** [rwth-i6/returnn-experiments#30]
I am trying to perform model adaptation for an encoder-decoder model (Similar to [full attention setup](https://github.com/rwth-i6/returnn-experiments/tree/master/2018-asr-attention/librispeech/full-setup-attention) ). For this, I'm freezing (trainable=False) all the layers except a couple of layers( let's say output_prob and readout ). I have 2 questions regarding this :- 

1. Is there any easy way to achieve this in the config. Right now I'm adding trainable=False for every layer except output_prob and readout and running the config.

2. I noticed that training speed was similar for my initial config and my config with only 2 layers trainable. I think that's because the gradient is being calculated for all layers. Is there any way to speed up the training given that I only need to train 2 layers and that too towards the end of the network. 

Thanks in advance for your answer.

**Comment by albertz** [comment]
> 1. Is there any easy way to achieve this in the config. Right now I'm adding trainable=False for every layer except output_prob and readout and running the config.

That's the right way.

In the output log, check that you see this:

`net trainable params: [<tf.Variable 'output/rec/output_prob/W' shape=(500, 1031) initial_value=<tf.Tensor 'output/rec/output_prob/W/Initializer/random_uniform:0' shape=(500, 1031) dtype=float32>>, <tf.Variable 'output/rec/output_prob/b' shape=(1031,) initial_value=<tf.Tensor 'output/rec/output_prob/b/Initializer/zeros:0' shape=(1031,) dtype=float32>>, <tf.Variable 'output/rec/readout_in/W' shape=(1024, 1000) initial_value=<tf.Tensor 'output/rec/readout_in/W/Initializer/random_uniform:0' shape=(1024, 1000) dtype=float32>>, <tf.Variable 'output/rec/readout_in/b' shape=(1000,) initial_value=<tf.Tensor 'output/rec/readout_in/b/Initializer/zeros:0' shape=(1000,) dtype=float32>>, <tf.Variable 'output/rec/s/rec/W' shape=(2030, 4096) initial_value=<tf.Tensor 'output/rec/s/rec/W/Initializer/random_uniform:0' shape=(2030, 4096) dtype=float32>>, <tf.Variable 'output/rec/s/rec/W_re' shape=(1024, 4096) initial_value=<tf.Tensor 'output/rec/s/rec/W_re/Initializer/random_uniform:0' shape=(1024, 4096) dtype=float32>>, <tf.Variable 'output/rec/s/rec/b' shape=(4096,) initial_value=<tf.Tensor 'output/rec/s/rec/b/Initializer/Const:0' shape=(4096,) dtype=float32>>]`

It lists all the trainable parameters. It should only list those which you marked as trainable. (My given example is similar as what you said, output_prob and readout, and also s.)

> 2. I noticed that training speed was similar for my initial config and my config with only 2 layers trainable. I think that's because the gradient is being calculated for all layers. Is there any way to speed up the training given that I only need to train 2 layers and that too towards the end of the network.

The gradient should only be calculated up to the point where it is needed. E.g. it should not be calculated for the encoder layers. So maybe you have some bug in the config still (check trainable params in the output). But it could also be that the gradient of output_prob dominates, such that the difference is not too big. How big is your vocab? For faster speed, it can help to reduce the BPE vocab (e.g. 1k). You could also do some [profiling](https://returnn.readthedocs.io/en/latest/profiling.html), and compare the output in each case. That should give you a better understanding about what part takes how long.

**Comment by abhigarg-iitk** [comment]
Thanks for your reply. I checked my returnn code. 

> It lists all the trainable parameters. It should only list those which you marked as trainable

It seems like trainable=False is only working if I put it on the output layer. If I apply trainable=False to each layer individually inside output, it still trains them.

This might be due to my version of returnn. Due to my personal reasons I'm using an older version of returnn. Is this a new feature that was added? Is it possible for you to point me to the corresponding commit? 

> The gradient should only be calculated up to the point where it is needed. E.g. it should not be calculated for the encoder layers.

Is this also a relatively new feature? Can you please also point me to it's commit?

**Comment by abhigarg-iitk** [comment]
I think the issue is resolved. It would be great if you can answer the above questions. Thanks, I'll close the issue.

**Comment by albertz** [comment]
> > It lists all the trainable parameters. It should only list those which you marked as trainable
> 
> It seems like trainable=False is only working if I put it on the output layer. If I apply trainable=False to each layer individually inside output, it still trains them.
> This might be due to my version of returnn. Due to my personal reasons I'm using an older version of returnn. I will merge [this](https://github.com/rwth-i6/returnn/pull/118/commits) commit.

Ah yes.
In general, I would always recommend that you keep using the latest version.
If there are problems in the latest version, we could just fix those, which would benefit everyone. I don't think that it's a good solution to stay at an older version because of some issues or so.

> > The gradient should only be calculated up to the point where it is needed. E.g. it should not be calculated for the encoder layers.
> 
> Is this also a relatively new feature? Can you please point me to it's commit?

This is nothing specific for Returnn. This is in general how TensorFlow calculates gradients (`tf.gradients`), or also how in general automatic differentiation works.

But as said, do some [profiling](https://returnn.readthedocs.io/en/latest/profiling.html), you should actually see where your computation time goes, and this would be helpful to understand how to optimize it.
